http/src/main/java/com/github/awsjavakit/http/ParameterStoreCachedTokenProvider.java

```diff
@@ -0,0 +1,83 @@
+package com.github.awsjavakit.http;
+
+import static com.gtihub.awsjavakit.attempt.Try.attempt;
+import static java.util.Objects.isNull;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.github.awsjavakit.http.token.OAuthTokenEntry;
+import java.time.Duration;
+import java.time.Instant;
+import software.amazon.awssdk.services.ssm.SsmClient;
+import software.amazon.awssdk.services.ssm.model.GetParameterRequest;
+import software.amazon.awssdk.services.ssm.model.GetParameterResponse;
+import software.amazon.awssdk.services.ssm.model.Parameter;
+import software.amazon.awssdk.services.ssm.model.PutParameterRequest;
+
+public class ParameterStoreCachedTokenProvider implements TokenProvider {
+
+  public static final String DATATYPE = "text";
+  public static final String TIER = "Standard";
+  private final TokenProvider newTokenProvider;
+  private final Duration maxTokenAge;
+  private final SsmClient ssmClient;
+  private final ObjectMapper objectMapper;
+  private final String parameterName;
+
+  public ParameterStoreCachedTokenProvider(TokenProvider newTokenProvider,
+    String ssmParameterName,
+    SsmClient ssmClient,
+    Duration maxTokenAge,
+    ObjectMapper objectMapper) {
+    this.ssmClient = ssmClient;
+    this.newTokenProvider = newTokenProvider;
+    this.maxTokenAge = maxTokenAge;
+    this.objectMapper = objectMapper;
+    this.parameterName = ssmParameterName;
+  }
+
+  @Override
+  public String fetchToken() {
+    var now = Instant.now();
+    var token = fetchTokenFromSsm();
+    if (thereIsNoValidToken(token, now)) {
+      var newValue = newTokenProvider.fetchToken();
+      token = storeNewToken(newValue);
+    }
+    return token.value();
+  }
+
+  private Boolean thereIsNoValidToken(OAuthTokenEntry token, Instant now) {
+    return isNull(token) || now.isAfter(token.timestamp().plus(maxTokenAge));
+  }
+
+  private OAuthTokenEntry fetchTokenFromSsm() {
+    GetParameterRequest request = GetParameterRequest.builder()
```

Comment:
You can make a weak warning for typed variables that can be vars in intellij btw